src/core/lib/iomgr/polling_interface.h

```diff
@@ -0,0 +1,167 @@
+  // from grpc_pollset_work before any events are received and before the
+  // timeout has expired. It is both initialized and destroyed by
+  // grpc_pollset_work. Initialization of worker is guaranteed to occur BEFORE
+  // the pollset's mutex is released for the first time by grpc_pollset_work and
+  // it is guaranteed that it will not be released by grpc_pollset_work AFTER
+  // worker has been destroyed.
+  //
+  // It's legal for worker to be NULL: in that case, this specific thread can
+  // not be directly woken with a kick, but maybe be indirectly (with a kick
+  // against the pollset as a whole).
+  //
+  // Tries not to block past deadline. May call grpc_closure_list_run on
+  // grpc_closure_list, without holding the pollset lock
+  virtual grpc_error* Poll(grpc_exec_ctx* exec_ctx, Worker** worker,
+                           grpc_millis deadline) GRPC_MUST_USE_RESULT = 0;
+
+  // Break one polling thread out of polling work for this pollset.
+  // If specific_worker is non-NULL, then kick that worker.
+  virtual grpc_error* Kick(grpc_exec_ctx* exec_ctx,
+                           Worker* specific_worker) GRPC_MUST_USE_RESULT = 0;
+
+  // Begin shutting down the pollset, and call on_done when done.
+  // pollset's mutex must be held
+  virtual void Shutdown(grpc_exec_ctx* exec_ctx, grpc_closure* on_done) = 0;
+
+  // CHANGING: return the size of a Poller with the current polling engine
+  // (will be moved to the polling engine interface directly in a future
+  // revision)
+  static size_t PollerSize();
+
+  // CHANGING: instantiate a Poller with the current polling engine
+  static Poller* Create(void* memory, gpr_mu** mu);
+
+ protected:
+  Poller() {}
+
+ private:
+  void AddToPollingJoin(PollingJoin* polling_join) override final;
+};
+
+// A PollingJoin serves to join multiple pollers and pollables.
+// Each Pollable that's added to a PollingJoin is added to ALL Poller instances
+// contained within it
+// Each Poller that's added to a PollingJoin receives ALL Pollable's that have
+// been added to it
+// (both in perpetuity)
+//
+// Using a PollingJoin is often cheaper than manually updating the same using
+// just the Poller/Pollable interfaces as some polling engines have short-cut
+// paths to bulk update pollers/pollables
+class PollingJoin : public PollableSet {
+ public:
+  // inherits AddPollable
+
+  // remove a pollable from a join - it's unspecified whether this stops
+  // existing Poller's in this join from polling on the Pollable, but guarantees
+  // it will not be added to future Pollables
+  virtual void RemovePollable(Pollable* pollable) = 0;
+
+  // Add a Poller to the join
+  virtual void AddPoller(Poller* poller) = 0;
+  // and remove it again (so it doesn't receive new Pollables)
+  virtual void RemovePoller(Poller* poller) = 0;
+
+  // Merge a PollingJoin with this one
+  // Each PollingJoin subsequently acts as a different handle to the same
```

Comment:
Let's explicitly document the exact semantics of a merge.

For example, consider the following case:
- create PollingJoin A
- create PollingJoin B
- merge A and B
- add Poller to A
- add Pollable to B

I assume that polling on the Poller added to A will then include the Pollable added to B?  And if the Pollable is subsequently removed from B, that polling on the Poller may no longer include the Pollable (subject to the same caveat documented for `RemovePollable()` above)?